### PR TITLE
Issue #117: Add Check API status report support for lock staleness

### LINKED ISSUE #117: Add Check API integration for stale locks
This could catch any locks which have been active for X period, pointing to a stale lock to cleanup, or a stuck task that has gotten stalled.

Inspired by finding a lock that was 27 days old, with the process still active on the cron host, just spinning its wheels.

### PR DESCRIPTION
**Pull request by Peterburnett**
Closes #117

### CODE REVIEW

File: lang/en/tool_lockstats.php [hunk 1]
```diff
@@ -46,12 +46,14 @@
...
 $string['h1_nexttask'] = 'Next running tasks';
 $string['lock_in_use'] = 'Lock still in use by another process';
+$string['nostalelocks'] = 'No stale process locks detected.';
 $string['pluginname'] = 'Cron Lock statistics';
 $string['privacy:metadata'] = 'Cron Lock statistics plugin does not store any personal data.';
 $string['release_all_locks'] = 'Release all stale locks';
 $string['release_lock'] = 'Release lock';
 $string['reset_header'] = 'Reset lock statistics history';
 $string['reset_text'] = 'Reset the lock stastics history.';
+$string['stalelocks'] = 'Stale process locks detected.';
```
Comment (dmitriim) [inline lang/en/tool_lockstats.php]: maybe should mention that locks are older that 24 hours?

File: lang/en/tool_lockstats.php [hunk 2]
```diff
@@ -46,12 +46,14 @@
 $string['h1_slowest'] = 'Slowest tasks this week with a duration > {$a} seconds';
 $string['h1_nexttask'] = 'Next running tasks';
 $string['lock_in_use'] = 'Lock still in use by another process';
+$string['nostalelocks'] = 'No stale process locks detected.';
```
Comment (dmitriim) [inline lang/en/tool_lockstats.php]: here as well would be good to mention that no locks older than 24 hours

Comment (Peterburnett) [reply lang/en/tool_lockstats.php]: good thought, I will add it

File: classes/check/stale_lock.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Form for resetting the history.
+ *
+ * @package    tool_lockstats
+ * @author     Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright  Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class stale_lock extends check {
```
Comment (dmitriim) [inline classes/check/stale_lock.php:29]: I'm not sure if there is a way, but would be good to add any sort of auto tests.

**Comment by brendanheywood** [comment]
This feels like an anti feature to me. Instead of reporting on this, why don't we just make a cron job which gracefully cleans them up?

**Comment by brendanheywood** [comment]
To clarify my misunderstanding:

1) Broadly I'm treating this plugin as deprecated and ideally new checks that make sense should be in core not here
2) The equivalent check for doing this in core is: https://tracker.moodle.org/browse/MDL-72775 (I'm doing this now)
3) What I thought this check was doing was looking for bad lock metadata, rather than actual locks. There is another core bug which also needs to be fixed so this doesn't produce false alarms:

https://tracker.moodle.org/browse/MDL-72268

https://tracker.moodle.org/browse/MDL-73293 (duplicate)

**Pull request merged**